src/dispense_tracker.rs

```diff
@@ -0,0 +1,96 @@
+use std::{
+    cmp::Ordering,
+    collections::{BinaryHeap, HashMap, HashSet},
+};
+
+use fuel_types::Address;
+
+#[derive(Debug, Eq, PartialEq)]
+pub(crate) struct Entry {
+    address: Address,
+    timestamp: u64,
+}
+
+impl Ord for Entry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        other.timestamp.cmp(&self.timestamp)
+    }
+}
+
+impl PartialOrd for Entry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+pub trait Clock: std::fmt::Debug + Send + Sync {
+    fn now(&self) -> u64;
+}
+
+#[derive(Debug)]
+pub struct TokioTime {}
+
+impl Clock for TokioTime {
+    fn now(&self) -> u64 {
+        tokio::time::Instant::now().elapsed().as_secs()
+    }
+}
+
+#[derive(Debug)]
+pub struct DispenseTracker {
+    tracked: HashMap<Address, u64>,
+    queue: BinaryHeap<Entry>,
+    in_progress: HashSet<Address>,
+    clock: Box<dyn Clock>,
+}
+
+impl Default for DispenseTracker {
+    fn default() -> Self {
+        Self {
+            tracked: Default::default(),
+            queue: Default::default(),
+            in_progress: Default::default(),
+            clock: Box::new(TokioTime {}),
+        }
+    }
+}
+
+impl DispenseTracker {
+    pub fn new(clock: impl Clock + 'static) -> Self {
+        Self {
+            tracked: HashMap::new(),
+            queue: BinaryHeap::new(),
+            in_progress: HashSet::new(),
+            clock: Box::new(clock),
+        }
+    }
+
+    pub fn track(&mut self, address: Address) {
+        self.in_progress.remove(&address);
+
+        let timestamp = self.clock.now();
+        self.tracked.insert(address, timestamp);
+        self.queue.push(Entry { address, timestamp });
+    }
+
+    pub fn mark_in_progress(&mut self, address: Address) {
+        self.in_progress.insert(address);
+    }
```

Comment:
You don't use it anyhow. Plus, you need to handle the case of the error and remove the user from `in_progress` collection.

Reply:
You're right, it was meant to be used inside `has_reached_dispense_limit` instead of `.track()`.

The problem I was trying to solve was this:
If someone requests assets, I need to track that as soon as possible so that concurrent requests don't bypass the limit. Using `track()` would be suboptimal since I need to `untrack()` on failure and that would require traversing the whole queue.
I assumed that handling the error case was negligible since the errors that can happen seem unlikely or would warrant a restart of the faucet anyways.

Reply:
I know what you are trying to fix, and I'm trying to say that you need to remove the user from `in_progress` collection on failure=) Because removing the `in_progress` from the current implementation doesn't change the flow.

BTW, we shouldn't restart the faucet on the error. If we do that, then we need to fix it.

Reply:
Got it. Thanks! 😄